File: src/array/SubarraySumEqualsK_560.java
package array;

import java.util.HashMap;
import java.util.Map;

/**
 * Given an array of integers and an integer k,
 * you need to find the total number of continuous subarrays whose sum equals to k.
 */
public class SubarraySumEqualsK_560 {

    /**
     * The length of the array is in range [1, 20,000].
     * The range of numbers in the array is [-1000, 1000] and the range of the integer k is [-1e7, 1e7].
     */
    public static void main(String[] args) {
        System.out.println(new SubarraySumEqualsK_560().subarraySum(new int[]{1, 1, 1}, 2));
        System.out.println(new SubarraySumEqualsK_560().subarraySum(new int[]{0, 0, 0, 0, 0, 0, 0, 0, 0, 0}, 0));
    }

    public int subarraySum(int[] nums, int k) {
        Map<Integer, Integer> cache = new HashMap<>();
        cache.put(0, 1);
        int res = 0;
        int sum = 0;
        for (int num : nums) {
            sum += num;

            if (cache.containsKey(sum - k))
                res += cache.get(sum - k);
            cache.put(sum, cache.getOrDefault(sum, 0) + 1);
        }
        return res;
    }
}

File: src/array/subarray_sum_equals_k_560.py
# sum += nums[i]
# while sum > k, subtract start from sum, increment start
# if sum == k, increment result, subtract start from sum, increment start

class Solution(object):
    def subarraySum2(self, nums, k):
        result = 0

        for i, num in enumerate(nums):
            sum = 0
            for j in range(i, len(nums)):
                sum += nums[j]
                if sum == k:
                    result += 1

        return result

    def subarraySum(self, nums, k):
        result = 0
        dict = {0: 1}
        sum = 0

        for num in nums:
            sum += num

            if sum - k in dict:
                result += dict[sum - k]

            dict[sum] = dict.get(sum, 0) + 1

        return result


print(Solution().subarraySum([-1, -1, 1], 1))
print(Solution().subarraySum([1, 1, 1], 2))
print(Solution().subarraySum([1, 2, 3], 3))
